### DISCUSSION TITLE
Feature request: rolling release

### DISCUSSION THREAD

**Discussion by Thijss** [networkit/networkit#1288] [Ideas]
Is it possible to switch to a rolling release system, so that e.g. every merge into `main` automatically results in a new release?
We could also use Semantic Versioning to indicate breaking changes etc.

The release process is currently quite slow and unpredictable in my experience.

I could help set this up if you are interested.

Cheers,

Thijs

**Comment by Thijss** [comment]
example of a project I'm familiar with, with rolling release system:

https://github.com/PowerGridModel/power-grid-model/
https://github.com/PowerGridModel/power-grid-model/releases

**Comment by fabratu** [comment]
Hi Thijss,

this sounds like a nice-to-have feature in general. We currently have something less useable/persistent in place. Wheels are builds and uploaded as artifacts for every merge to master. 

From your linked repo, it seems, that you use labels to mark release as "pre-release" and everything is pushed to the original repo. I would rather have all pre-releases pushed to an additional repo like https://github.com/networkit/nightly-builds to avoid creating too many entries in the main repo.

I guess there is a marketplace action in place, which handles the release pushes + artifact upload? In this case, do you know whether the behavior as described above is also possible?

**Comment by Thijss** [comment]
Hi Fabian,
Sorry for the late reply. I didn't realize you had replied and I forgot about this issue.

Personally, I don't think the behavior you describe above is desirable. Pushing nightly-builds/alplha/dev builds to a separate repository would make things unnecessary complex.

> ..avoid creating too many entries in the main repo.

I assume you mean 'releases' here. What would be the downside of having many releases in your opinion? 
In my opinion, the goal is to have more releases. In a rolling release system, every push to master would be a new **stable** release. 

Additionally, I would suggest switching to the SemVer versioning system (https://semver.org/). Then it will be clear which release are bug-fixes, which releases add features, and which releases contain breaking changes.

In summary, I still think this could be a nice feature to add, but it depends on how much faith there is in the current tests/CICD. If you trust the tests/CICD, a rolling release system would allow users to immediately benefit from merged features.

**Comment by Thijss** [comment]
For testing purposes, nightly/alpha/dev releases could also be created, but I think it would be wise to only upload stable releases to pypi. 

To avoid cluttering the repo with those, a cleanup action could be added that runs periodically or after each stable release.
